### PR TITLE
Remove -rdynamic from the linker.

### PR DESCRIPTION
**Pull request by Pentarctagon**
This PR:

- Adds the RemoveFlag() function.
- Replaces the existing logic removing the `-DNDEBUG` flag with RemoveFlag().
- Removes the [-rdynamic](https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html) flag from the linker.  Cmake adds this by default while scons does not, and it adds close to 5 MBs to the executable.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,20 @@ set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS true)
 # use our own version of FindBoost.cmake and other Find* scripts
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
 
+# function to remove a flag from a variable
+function(RemoveFlag VAR SCOPE FLAG)
+	MESSAGE("Removing ${FLAG} flag from ${VAR}")
+	separate_arguments(${VAR})
+	list(REMOVE_ITEM ${VAR} ${FLAG})
+	string(REPLACE ";" " " ${VAR} "${${VAR}}")
```
Comment (jyrkive) [inline CMakeLists.txt]: I'm not sure if the indirection is needed for the input. Try

```cmake
string(REPLACE ";" " " ${VAR} ${VAR})
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,6 +18,20 @@ set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS true)
...
+	
+	if("${SCOPE}" STREQUAL "CACHE")
+		set(${VAR} "${${VAR}}" CACHE STRING "" FORCE)
```
Comment (jyrkive) [inline CMakeLists.txt]: Likewise, here I'm not sure if the indirection is needed.

In addition, I'd like if you provided the `docstring` parameter. For these parameters, something like

> Default C++ flags for RelWithDebInfo
> Defaut C flags for RelWithDebInfo
> Default C++ flags for MinSizeRel
> Default C flags for MinSizeRel

File: CMakeLists.txt [hunk 3]
```diff
@@ -18,6 +18,20 @@ set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS true)
...
+	elseif("${SCOPE}" STREQUAL "PARENT")
```
Comment (jyrkive) [inline CMakeLists.txt]: This scope is the local scope from the POV of the caller. I'd prefer if you called this scope "LOCAL" or something.

File: CMakeLists.txt [hunk 4]
```diff
@@ -18,6 +18,20 @@ set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS true)
...
+		set(${VAR} "${${VAR}}" PARENT_SCOPE)
```
Comment (jyrkive) [inline CMakeLists.txt]: Likewise, indirection for the value may be unnecessary here.

File: CMakeLists.txt (hunk 1 above)
Comment (Pentarctagon) [reply CMakeLists.txt]: I tested to be sure, but it is necessary to do it that way, like how the `check_compiler_has_flag` macro does it.

File: CMakeLists.txt (hunk 2 above)
Comment (Pentarctagon) [reply CMakeLists.txt]: I could say "Removed ${FLAG}" in the docstring, but to have more would require either passing in more information as arguments or attempting to parse the variable name.  C vs C++ flags and the build being used should be fairly obvious from the variable name itself though, if/when someone does open up the CMakeCache.

File: CMakeLists.txt (hunk 3 above)
Comment (Pentarctagon) [reply CMakeLists.txt]: What about this?
```
if("${SET_IN}" STREQUAL "CACHE")
	set(${VAR} "${${VAR}}" CACHE STRING "Removed ${FLAG}" FORCE)
elseif("${SET_IN}" STREQUAL "SCRIPT")
	set(${VAR} "${${VAR}}" PARENT_SCOPE)
endif()
```

File: CMakeLists.txt (hunk 2 above)
Comment (jyrkive) [reply CMakeLists.txt]: Passing the docstring as an argument is what I'd prefer.

And the docstring is supposed to mean what the variable **is**, not what we have done to it.

File: CMakeLists.txt (hunk 3 above)
Comment (jyrkive) [reply CMakeLists.txt]: "SCRIPT" would be fine for me.

File: CMakeLists.txt [hunk 5]
```diff
@@ -19,15 +19,15 @@ set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS true)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
 
 # function to remove a flag from a variable
-function(RemoveFlag VAR SCOPE FLAG)
```
Comment (jyrkive) [inline CMakeLists.txt]: I think SCOPE was a better name for this parameter.

**Comment by gfgtdf** [comment]
The online documentation says that option is needed to generate stacktraces , does it mean that users can now no longer generate stacktraces while they could previously.

**Comment by jyrkive** [comment]
@gfgtdf Please don't quote it out of context.

> This option is needed for some uses of `dlopen` or to allow obtaining backtraces from within a program.

So, it would be needed to generate stacktraces *from within Wesnoth*. It doesn't affect external debuggers such as GDB.

**Comment by Pentarctagon** [comment]
For example, if you were using something like backtrace()/backtrace_symbols(), you would need `-rdynamic`.

**Comment by GregoryLundberg** [comment]
Pulled and resolved the merge conflict. Comparing executable sizes:

**BEFORE**
```
-rwxr-xr-x 1 lundberg users  23222736 Nov 21 10:16 ./boost_unit_tests
-rwxr-xr-x 1 lundberg users   1487304 Nov 21 10:14 ./campaignd
-rwxr-xr-x 1 lundberg users  29792096 Nov 21 10:52 ./wesnoth
-rwxr-xr-x 1 lundberg users   2947040 Nov 21 10:52 ./wesnothd
```
**AFTER**
```
-rwxr-xr-x 1 lundberg users  23222736 Nov 21 10:16 ./boost_unit_tests
-rwxr-xr-x 1 lundberg users   1487304 Nov 21 10:14 ./campaignd
-rwxr-xr-x 1 lundberg users  25532256 Nov 21 10:51 ./wesnoth
-rwxr-xr-x 1 lundberg users   2365408 Nov 21 10:51 ./wesnothd
```

These results are repeatable and stable.

I notice, however, that only wesnoth and wesnothd are re-linked and effected. I would expect boost_unit_tests and campaignd to be re-linked and effected as well.

**Comment by Pentarctagon** [comment]
I don't know what's going on in your case then.  This is what I see from cmake with/without `-rdynamic` and from scons:
```
2021432 campaignd-cmake-no-rdynamic
2697272 campaignd-cmake-rdynamic
2021432 campaignd-scons
```
The sizes generally don't match what you have, but the more important thing I was looking at with this is that the size of the cmake and scons builds match.

**Comment by GregoryLundberg** [comment]
OK. I'll retest from clean environments using defaults for scons and cmake, only adding tests and servers and comparing before/after on cmake.

**Comment by GregoryLundberg** [comment]
Oh and the comment about boost_unit_tests and campaignd was because when I switched back and forth the only activity was linking wesnoth and wesnothd. Repeated several times and no change, just those two re-linked.

**Comment by GregoryLundberg** [comment]
OK. Clean environments. Defaults except for also compiling boost unit tests and the servers.

```
/home/lundberg/Pentarctagon/cmake-all-release-g++:
total 65160
drwxr-xr-x 2 lundberg users     4096 Nov 21 13:06 .
drwxr-xr-x 5 lundberg users     4096 Nov 21 13:21 ..
-rwxr-xr-x 1 lundberg users 31449088 Nov 21 12:39 boost_unit_tests
-rwxr-xr-x 1 lundberg users  2519688 Nov 21 12:32 campaignd
-rwxr-xr-x 1 lundberg users 29792096 Nov 21 12:38 wesnoth
-rwxr-xr-x 1 lundberg users  2947040 Nov 21 12:32 wesnothd

/home/lundberg/Pentarctagon/cmake-all-release-g++-no-rdynamic:
total 55640
drwxr-xr-x 2 lundberg users     4096 Nov 21 13:22 .
drwxr-xr-x 5 lundberg users     4096 Nov 21 13:21 ..
-rwxr-xr-x 1 lundberg users 27062272 Nov 21 13:18 boost_unit_tests
-rwxr-xr-x 1 lundberg users  1999496 Nov 21 13:14 campaignd
-rwxr-xr-x 1 lundberg users 25532256 Nov 21 13:17 wesnoth
-rwxr-xr-x 1 lundberg users  2365408 Nov 21 13:14 wesnothd

/home/lundberg/Pentarctagon/scons-all-release-g++:
total 55600
drwxr-xr-x 2 lundberg users     4096 Nov 21 12:27 .
drwxr-xr-x 5 lundberg users     4096 Nov 21 13:21 ..
-rwxr-xr-x 1 lundberg users 27054080 Nov 21 12:17 boost_unit_tests
-rwxr-xr-x 1 lundberg users  1987208 Nov 21 12:16 campaignd
-rwxr-xr-x 1 lundberg users 25524064 Nov 21 12:16 wesnoth
-rwxr-xr-x 1 lundberg users  2353120 Nov 21 12:16 wesnothd
```

Tested switching back-and-forth between master and PR and, yes, it re-links all four executables, repeatable at the above sizes.

So, yes, this does reduce CMake to be nearly equal to SCons for the all defaults on GCC 7.2.0

If you correct the merge conflict, I'd suggest this should merge.

**Comment by jyrkive** [comment]
Manually cherry-picked to master as a4a1cec6373433bd88616bc239bcad72541ffa2a.

**Pull request closed without merging**